### PR TITLE
Java Build Aspects extractors

### PR DESCRIPTION
**Pull request by St33lPhoenix**
Add extractor for Maven compiler plugin configuration Add extractor for Java module detection
Add extractor for Java version extraction
Add extractor for parent artifacts
Refactor existing code

### CODE REVIEW

File: analyzer/src/main/java/nl/tudelft/mavensecrets/extractors/JavaVersionExtractor.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            result[1] = null;
+            result[2] = null;
+        } else {
+            try (InputStream stream = jar.getInputStream(entry)) {
+                // https://docs.oracle.com/javase/specs/jvms/se17/html/jvms-4.html#jvms-4.1
+
+                byte[] buf = new byte[8];
+
+                // Sanity check
+                if (stream.read(buf) != buf.length || buf[0] != 0xCA || buf[1] != 0xFE || buf[2] != 0xBA || buf[3] != 0xBE) {
```
Comment (cryocz) [inline analyzer/src/main/java/nl/tudelft/mavensecrets/extractors/JavaVersionExtractor.java]: Bytes are sign extended and therefore do not match

File: analyzer/src/main/java/nl/tudelft/mavensecrets/extractors/JavaVersionExtractor.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+/**
+ * An extractor fetching Java versions from an artifact.
+ * It looks both at <code>META-INF/MANIFEST.MD</code> and class bytecode.
+ * It may be possible the bytecode read is from a class shaded in rather than a class from the actual project, in which case a different version may be detected. 
+ */
+public class JavaVersionExtractor implements Extractor {
+
+    private final Field[] fields = {
+            new Field("java_version_manifest", "VARCHAR(16)"),
+            new Field("java_version_class_major", "BIT(16)"),
```
Comment (cryocz) [inline analyzer/src/main/java/nl/tudelft/mavensecrets/extractors/JavaVersionExtractor.java]: Postgres cannot coerce a byte array into `BIT(n)` apparently

**Review by cryocz — CHANGES_REQUESTED** [review]

File: analyzer/src/main/java/nl/tudelft/mavensecrets/extractors/JavaVersionExtractor.java (hunk 2 above)
Comment (St33lPhoenix) [reply analyzer/src/main/java/nl/tudelft/mavensecrets/extractors/JavaVersionExtractor.java]: What is the correct type for fixed-size binary data then? PostgreSQL dialect does not support `BINARY(n)`.

File: analyzer/src/main/java/nl/tudelft/mavensecrets/extractors/JavaVersionExtractor.java (hunk 1 above)
Comment (St33lPhoenix) [reply analyzer/src/main/java/nl/tudelft/mavensecrets/extractors/JavaVersionExtractor.java]: See latest commit.

File: analyzer/src/main/java/nl/tudelft/mavensecrets/extractors/JavaVersionExtractor.java (hunk 2 above)
Comment (cryocz) [reply analyzer/src/main/java/nl/tudelft/mavensecrets/extractors/JavaVersionExtractor.java]: Yeaaah, you either have to convert it into an `INT32` or use `BYTEA`

Comment (St33lPhoenix) [reply analyzer/src/main/java/nl/tudelft/mavensecrets/extractors/JavaVersionExtractor.java]: See latest commit.

**Review by cryocz — APPROVED** [review]
Looks good to me

**Comment by Vel1khan** [comment]
LGTM

**Pull request merged**